compiler/noirc_frontend/src/elaborator/enums.rs

```diff
@@ -243,4 +253,672 @@ impl Elaborator<'_> {
             (pattern, parameter_type, Visibility::Private)
         }))
     }
+
+    /// To elaborate the rules of a match we need to go through the pattern first to define all
+    /// the variables within, then compile the corresponding branch. For each branch we do this
+    /// way we'll need to keep a distinct scope so that branches cannot access the pattern
+    /// variables from other branches.
+    ///
+    /// Returns (rows, result type) where rows is a pattern matrix used to compile the
+    /// match into a decision tree.
+    pub(super) fn elaborate_match_rules(
+        &mut self,
+        variable_to_match: DefinitionId,
+        rules: Vec<(Expression, Expression)>,
+    ) -> (Vec<Row>, Type) {
+        let result_type = self.interner.next_type_variable();
+        let expected_pattern_type = self.interner.definition_type(variable_to_match);
+
+        let rows = vecmap(rules, |(pattern, branch)| {
+            self.push_scope();
+            let pattern = self.expression_to_pattern(pattern, &expected_pattern_type);
+            let columns = vec![Column::new(variable_to_match, pattern)];
+
+            let guard = None;
+            let body_span = branch.span;
+            let (body, body_type) = self.elaborate_expression(branch);
+
+            self.unify(&body_type, &result_type, || TypeCheckError::TypeMismatch {
+                expected_typ: result_type.to_string(),
+                expr_typ: body_type.to_string(),
+                expr_span: body_span,
+            });
+
+            self.pop_scope();
+            Row::new(columns, guard, body)
+        });
+        (rows, result_type)
+    }
+
+    /// Convert an expression into a Pattern, defining any variables within.
+    fn expression_to_pattern(&mut self, expression: Expression, expected_type: &Type) -> Pattern {
+        let expr_span = expression.span;
+        let unify_with_expected_type = |this: &mut Self, actual| {
+            this.unify(actual, expected_type, || TypeCheckError::TypeMismatch {
+                expected_typ: expected_type.to_string(),
+                expr_typ: actual.to_string(),
+                expr_span,
+            });
+        };
+
+        match expression.kind {
+            ExpressionKind::Literal(Literal::Integer(x, negative)) => {
+                let Some(mut x): Option<i64> = x.try_to_u64().and_then(|x| x.try_into().ok())
```

Comment:
Is there a reason why `Pattern::Int` takes an i64 instead of `(FieldElement, bool /* negative */)`? For example this code triggers the panic below:

```noir
fn main() {
    let x = 1;
    match x {
        18446744073709551615 => {
            println("yes")
        }
        _ => println("no")
    };
}
```

Reply:
(not blocking, though, as this scenario would be rare, but a follow-up issue could be captured)

Reply:
I didn't think I could get the range constructor working with the boolean. I tried it again though and it wasn't too bad - just had to create a wrapper type for `SignedField` and use that. This also fixed a bug with representation of signed fields in an enum tag value.